File: src/frontend/org/voltdb/exceptions/ConstraintFailureException.java
/* This file is part of VoltDB.
 * Copyright (C) 2008-2022 Volt Active Data Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with VoltDB.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.voltdb.exceptions;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.logging.Logger;

import org.voltdb.PrivateVoltTableFactory;
import org.voltdb.VoltDB;
import org.voltdb.VoltTable;
import org.voltdb.client.ClientResponse;
import org.voltdb.messaging.FastDeserializer;
import org.voltdb.types.ConstraintType;

/**
 * Exception generated when a constraint is violated. Contains more information then a SQLException.
 *
 */
public class ConstraintFailureException extends SQLException {

    protected static final Logger LOG = Logger.getLogger(ConstraintFailureException.class.getName());

    /**
     * Constructor for deserializing an exception returned from the EE.
     * @param exceptionBuffer
     */
    public ConstraintFailureException(ByteBuffer exceptionBuffer) {
        super(exceptionBuffer);
        type = ConstraintType.get(exceptionBuffer.getInt());
        try {
            tableName = FastDeserializer.readString(exceptionBuffer);
        }
        catch (IOException e) {
            // implies that the EE created an invalid constraint
            // failure, which would be a corruption/defect.
            VoltDB.crashLocalVoltDB(e.getMessage(), true, e);
        }
        if (exceptionBuffer.hasRemaining()) {
            int tableSize = exceptionBuffer.getInt();
            buffer = ByteBuffer.allocate(tableSize);
            //Copy the exception details.
            exceptionBuffer.get(buffer.array());
        } else {
            buffer = ByteBuffer.allocate(0);
        }
    }

    /**
     * Get the type of constraint violation that occurred
     * @return Type of constraint violation
     */
    public ConstraintType getType() {
        return type;
    }

    public String getTableName() {
        return tableName;
    }

    /**
     * Get the table containing the involved tuples (target and source) that caused the constraint violation.
     */
    public VoltTable getTuples() {
        if (table != null) {
            return table;
        }

        table = PrivateVoltTableFactory.createVoltTableFromSharedBuffer(buffer);

        return table;
    }

    /**
     * ByteBuffer containing a serialized copy of a table with the tuples involved in the constraint violation
     */
    private final ByteBuffer buffer;

    /**
     * Type of constraint violation that caused this exception to be thrown
     */
    private final ConstraintType type;

    /**
     * Name of the table that the constraint violation occurred in.
     */
    private String tableName = null;

    /**
     * Lazy deserialized copy of a table with the tuples involved in the constraint violation
     */
    private VoltTable table = null;

    @Override
    public String getMessage() {
        if (buffer.capacity() == 0) {
            return super.getMessage();
        } else {
            StringBuilder sb = new StringBuilder(super.getMessage());
            sb.append('\n');
            sb.append("Constraint Type ");
            sb.append(type);
            sb.append(", Table CatalogId ");
            sb.append(tableName);
            sb.append('\n');
            sb.append("Relevant Tuples:\n");
            sb.append(getTuples().toFormattedString());
            // remove trailing whitespace
            while (Character.isWhitespace(sb.charAt(sb.length() - 1))) {
                sb.setLength(sb.length() - 1);
            }
            return sb.toString();
        }
    }
    /**
     * Retrieve a string representation of the exception. If this is a rethrown HSQLDB exception it returns the enclosed exceptions
     * string. Otherwise a new string is generated with the details of the constraint violation.
     */
    @Override
    public String toString() {
        return getMessage();
    }

    @Override
    protected SerializableExceptions getExceptionType() {
        return SerializableExceptions.ConstraintFailureException;
    }

    @Override
    public byte getClientResponseStatus() {
        return ClientResponse.GRACEFUL_FAILURE;
    }

    @Override
    public String getShortStatusString() {
        return "CONSTRAINT VIOLATION";
    }

    @Override
    protected int p_getSerializedSize() {
        // ... + 8 + string prefix + string length + ...
        return super.p_getSerializedSize()
            + 4 // constraint type
            + 4 // table name string length
            + tableName.length()
            + 4 // buffer length
            + buffer.capacity();
    }

    @Override
    protected void p_serializeToBuffer(ByteBuffer b) {
        super.p_serializeToBuffer(b);
        b.putInt(type.getValue());
        b.putInt(tableName.length());
        b.put(tableName.getBytes());
        b.putInt(buffer.capacity());
        buffer.rewind();
        b.put(buffer);
    }

    /**
     *
     */
    private static final long serialVersionUID = 1L;

}

File: src/ee/storage/ConstraintFailureException.cpp
/* This file is part of VoltDB.
 * Copyright (C) 2008-2022 Volt Active Data Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with VoltDB.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "storage/ConstraintFailureException.h"
#include "storage/constraintutil.h"
#include "storage/table.h"
#include <common/debuglog.h>

using namespace voltdb;
using std::string;

ConstraintFailureException::ConstraintFailureException(
        Table *table, TableTuple tuple, TableTuple otherTuple,
        ConstraintType type, PersistentTableSurgeon *surgeon) :
    SQLException(SQLException::integrity_constraint_violation,
            "Attempted violation of constraint",
            VoltEEExceptionType::VOLT_EE_EXCEPTION_TYPE_CONSTRAINT_VIOLATION),
    m_table(table),
    m_tuple(tuple),
    m_otherTuple(otherTuple),
    m_type(type),
    m_surgeon(surgeon) {
    vassert(table);
    vassert(!tuple.isNullTuple());
}

ConstraintFailureException::ConstraintFailureException(
        Table *table,
        TableTuple tuple,
        string const& message,
        PersistentTableSurgeon *surgeon) :
        SQLException(
                SQLException::integrity_constraint_violation,
                message,
                VoltEEExceptionType::VOLT_EE_EXCEPTION_TYPE_CONSTRAINT_VIOLATION),
    m_table(table),
    m_tuple(tuple),
    m_otherTuple(TableTuple()),
    m_type(CONSTRAINT_TYPE_PARTITIONING),
    m_surgeon(surgeon) {
    vassert(table);
    vassert(!tuple.isNullTuple());
}

void ConstraintFailureException::p_serialize(ReferenceSerializeOutput *output) const {
    SQLException::p_serialize(output);
    output->writeInt(m_type);
    output->writeTextString(m_table->name());
    size_t tableSizePosition = output->reserveBytes(4);
    TableTuple tuples[] = { m_tuple, m_otherTuple };
    if (m_otherTuple.isNullTuple()) {
        m_table->serializeTupleTo(*output, tuples, 1);
    } else {
        m_table->serializeTupleTo(*output, tuples, 2);
    }
    output->writeIntAt(tableSizePosition, static_cast<int32_t>(output->position() - tableSizePosition - 4));
}

ConstraintFailureException::~ConstraintFailureException() throw () {
    // if delayed tuple deallocation for serialization (by passing in tableSurgeon),
    // do cleanup here
    VOLT_DEBUG("ConstraintFailureException has table surgeon %s", ((m_surgeon!=NULL) ? "true": "false"));
    if (m_surgeon && !m_tuple.isNullTuple()) {
        m_surgeon->deleteTupleStorage(m_tuple);
    }
}

string ConstraintFailureException::message() const {
    // This should probably be an override of the << operator and then used here, but meh
    string msg = SQLException::message();
    msg.append("\nConstraint violation type: ");
    string type_string = constraintutil::getTypeName(m_type);
    msg.append(type_string);
    msg.append("\non table: ");
    msg.append(m_table->name());

    msg.append("\nNew tuple:\n\t");
    msg.append(m_tuple.debug(m_table->name()));
    if (!m_otherTuple.isNullTuple()) {
        msg.append("\nOriginal tuple:\n\t");
        msg.append(m_otherTuple.debug(m_table->name()));
    }
    msg.append("\n");
    return msg;
}
